### DISCUSSION TITLE
loading files via go 1.16 embed directives

### DISCUSSION THREAD

**Discussion by yassinm** [cue-lang/cue#1145] [Q&A] [answered]
related: #1117

I tried loading files via go embed directives. I am actually not able to make this work. Here is the [playground link](https://play.golang.org/p/HF5cBu2f_jt) i was able to make. The playground is not working now because it is timing out. 

What i am trying to do is load a start file from the local file system than depends on files that are shipped with the binary as embed in go 1.16. main.cue should be able to reference pipeline.Schema

**Comment by roylisto** [comment]
Have you tried this?
`load.Instances([]string{}, &load.Config{Dir: "/path/to/cue/files"})`

**Comment by myitcv** [comment]
@yassinm please can you provide a reproduction of what is not working?

**Reply by yassinm** [reply]
@roylisto I am able to load the main.cue file from the file system but it is not able to find the pipeline package. That was my first try. Then i tried to put both files in the overlay so they get loaded by doing  `load.Instances([]string{"/" + Root + "/scripts/..."}, config)` as below. In the first case it is not able to find the pipeline package. In the second its not even loading main.cue

```
config := &load.Config{
		// Module:  Root,
		Dir:     "/",
		Overlay: overlay,
	}

	// insts := load.Instances([]string{ovpath}, config)
	insts := load.Instances([]string{"/" + Root + "/scripts/..."}, config)
```

```
-- go.mod --
module play.ground
-- scripts/main.cue --
package main

import (
  "play.ground/scripts/pipeline"
)

s : pipeline.#Schema & {
  Name: "jdoe"
  Age: 2000
  bebe: 11
}


-- scripts/pipeline/schema.cue --

package pipeline

#Schema: {
	Name: string
	Age:  >1000
	bebe: >1
}
```

@myitcv  I was hoping to have a simple reproducible example by using go playground (which is the link i provided above) that people can just run and see what is happening. Unfortuntely that is not working because building the simple case takes too long and go playground is barking at me. If it makes it easy i can create a repo with the reproducible code. Would that help ?

**Reply by myitcv** [reply]
@yassinm 

> If it makes it easy i can create a repo with the reproducible code. Would that help ?

Yes please! Please feel free to create a repro however you like. The advice in https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers might give you some ideas, but please don't feel you have to follow any approach listed there. The main thing is that we get a repro that I can consistently use to get the same results as you.

**Reply by yassinm** [reply]
@myitcv Good news !

i was able to reproduce the issue i am having in this [repo branch ](https://github.com/yassinm/cue-tests/tree/rc01)

The bad news is that i have [another branch](https://github.com/yassinm/cue-tests/tree/rc00) that is doing what i was expecting to happen. Meaning no failures. The only difference between these two branches is that i introduced more packages in my go code. I cant tell why that makes any difference so maybe someone can help here :-)

**Reply by verdverm** [reply]
There are more differences in your example branches

- the `//go:embed` directives are different. Maybe that is the cause?
- Path construction in `loadFiles` is different, try printing what the overlay and FS look like

**Reply by ymolists** [reply]
Those are some minor differences. I am actully printing the resulting overlay path  in both cases and they are the same.

```
//failing rc01
starting ... 
dpath  pipeline/schema.cue
ovpath  /play.ground/scripts/pipeline/schema.cue
_|_ // import failed: cannot find package "play.ground/scripts/pipeline"


//working rc00
starting ... 
dpath  scripts/pipeline/schema.cue
ovpath  /play.ground/scripts/pipeline/schema.cue
{
	s: {
		Name: "jdoe"
		Age:  2000
		bebe: 11
	}
}
```

What i am suspecting is that in rc00 its prolly just reading from the file system and ignoring the overlay. Where as in rc01 there is no way it can fall back to the file system so it ends up failing. Long story short i have to figure how to make it work with the overlay.

**Reply by verdverm** [reply]
I'd run the programs from within an empty directory. This way, as you noted, it cannot fallback on the filesystem.

the `dpath` is different above ie. the `scripts` part is missing in the non-working example

You aren't using a `cue.mod` directory, is that correct? Rather the repo module is called `play.ground`?

**Reply by ymolists** [reply]
I ran rc00 from the local dir as you suggested and sure enough it fails with the same error. I update the `app.go` and vscode lauch configuration in the repo to show what i had to do to make rc00 branch fail. So that is now consistent behaviour as i was suspecting.

dpath is not important. dpath is where the file is loaded from in the filesystem. but after that we store it i nthe overlay fs. As you can see the overlay fs path (logged as ovpath above) are the same.

```
	config := &load.Config{
		Module:  Root,
		Overlay: overlay,
	}
```
This is what i pass to the `load.Instances` call. I suspect there are more fields i could fill in the config type but i cant figure that for now. I do fill the Module field which i suspect tells load which module to use.

Whats the correct way to use an overlay ? is there an example that show this in the documentation ? I think this is the crux of the issue.

**Reply by verdverm** [reply]
I think you need to setup your project as a cue module. How does cue know where to find the import?

try `cue eval main.cue` on the following

```
-- foo/bar.cue --
package foo

A: "bar"
-- main.cue --
package main

import "play.cloud/foo"

a: foo.A
```

vs

```
-- cue.mod/module.cue --
module: "play.cloud"
-- foo/bar.cue --
package foo

A: "bar"
-- main.cue --
package main

import "play.cloud/foo"

a: foo.A
```

**Reply by ymolists** [reply]
I am sure the cue file system you suggested works. But that is not the issue.

I guess the real question we are trying to find here is about using overlay. How does cue find an import when using an overlay fs ? what do we have to pass to the load api in terms of configuration ? do we have to add a module file in the overlay fs as well ? i have not seen that documented so maybe i missed it.

**Reply by verdverm** [reply]
You need a `cue.mod/module.cue` file, overlay simply adds to the files `cue` sees. There is no module signifier, therefore you will not be able to use imports. That is what I suspect the error is

**Reply by ymolists** [reply]
@verdverm i want to say that i really appreciate your help here and for bearing with me :-)

I added this file to the overlay fs:
```
-- cue.mod/module.cue --
module: "play.ground"
```

This is what is outputed when its being loaded. notice how its path is ending up being `/play.ground/cue.mod/module.cue`.
```
[updated]
starting ... 
dpath  pipeline/module.cue
ovpath  /play.ground/cue.mod/module.cue

dpath  pipeline/schema.cue
ovpath  /play.ground/scripts/pipeline/schema.cue

_|_ // import failed: cannot find package "play.ground/scripts/pipeline"


```

I added the code to [the rc01 branch](https://github.com/yassinm/cue-tests/tree/rc01) i am still thinking the module has to be defined via the fields passed to the Config type during loading but i could be wrong or missing something. It was definitely worth a shot :-)

**Reply by verdverm** [reply]
The `ovpath` has 2 scripts in it, pathing is off

**Reply by ymolists** [reply]
Correct me if i am wrong but i dont think the location in the file system should matter.  Meaning the file is loaded into memory in a map. 

Can you confirm its ending up in the correct location in the overlay ?

**Reply by ymolists** [reply]
@verdverm  Mea culpa

I read your comment as two script files are in the file system.
The code in the repo was correct. What i pasted was wrong. I edited the comment above to reflect what i am actually getting.

**Reply by verdverm** [reply]
You may need to set some of the path variables in the Config object passed to load. You'll have to deal with the absolute path of the overlay and the CWD path of the first arg passed to the load function. The absolute path is outside of the working directory. You could deal with it by making everything absolute or relative to CWD. I'd go absolute so that you don't overlap with files on disk. Using a temp dir is recommended

**Reply by ymolists** [reply]
@verdverm thank you so much for all your help ! I was able to figure it based on what you mentioned.

This [test code](https://github.com/cue-lang/cue/blob/06484a39d8d44c656212ac2d3b5589f8aa9db033/cue/load/loader_test.go#L355) was also quite helpful. I am leaving what i did in here for posterity. I might remove that repo at some point. Basically we are tricking cue in loading a file from memory while it is mapped to a path that looks like a file from the local file system. As you can see cue.mod is also not necessary. I don't know how that works but it does work . That is for maybe another day ! Cheers !

```go
main.go


const (
	Root = "play.ground"
)

func Run() error {
	cwd, err := os.Getwd()
	if err != nil {
		return err
	}

	overlay, err := loadFiles(scripts.StaticFs, cwd)
	if err != nil {
		return err
	}

	config := &load.Config{
		Module:  Root,
		Overlay: overlay,
	}

	insts := load.Instances([]string{"./local/main.cue"}, config)

	ctx := cuecontext.New()
	for _, instIdx := range insts {
		inst := ctx.BuildInstance(instIdx)

		val := inst.Value()
		fmt.Printf("%v\n", val)
	}

	return nil
}

func loadFiles(fsys fs.FS, rDir string) (map[string]load.Source, error) {
	overlay := make(map[string]load.Source)

	abs := func(dpath string) string {
		return path.Join(rDir, dpath)
	}

	err := fs.WalkDir(fsys, ".", func(dpath string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if d == nil || d.Type().IsDir() || !d.Type().IsRegular() {
			return nil
		}

		ovpath := abs(path.Join("scripts", dpath))
		if !strings.HasSuffix(dpath, ".cue") {
			return nil
		}

		f, err := fsys.Open(dpath)
		if err != nil {
			return err
		}

		buf, err := io.ReadAll(f)
		if err != nil {
			return err
		}

		fmt.Println("dpath ", dpath)
		fmt.Println("ovpath ", ovpath)
		fmt.Println("")

		overlay[ovpath] = load.FromBytes(buf)

		return nil
	})

	if err != nil {
		return nil, err
	}

	return overlay, nil
}

-- scripts/main.cue --
package main

import (
  "play.ground/scripts/pipeline"
)

s : pipeline.#Schema & {
  Name: "jdoe"
  Age: 2000
  bebe: 11
}


-- scripts/pipeline/schema.cue --

package pipeline

#Schema: {
	Name: string
	Age:  >1000
	bebe: >1
}

```

**Comment by myitcv** [comment, marked as answer]
https://github.com/cue-lang/cue/issues/607#issuecomment-873389451 is relevant here, specifically:

> Put another way, an overlay, in its current form, exists in order to _complement_ what is on disk, rather than wholesale replace its contents.

The comment goes on to suggest how we will like add support for loading from an `io/fs.FS` in the future, via a `cuelang.org/go/cue/load.Config` field or similar.

Until that time, you need to follow an approach that, as @verdverm points out above, runs from an empty directory to be sure that the load does not interact with files on disk. This effectively becomes equivalent to writing the your `io/fs.FS` to a temporary directory FWIW. As @verdverm also mentioned, a `cue.mod/module.cue` will be required in the FS for non-builtin imports to work.

Hence I have the following example, which happens to use an overlay (but could just as easily write the `io/fs.FS` files to the temporary directory and dispense with the overlay). The goal of this example is to load the configuration that corresponds to the `main` CUE package, i.e. `scripts/main.cue`, from the `go:embed` FS, and then print that value:

```
# Populate our go.sum
go mod tidy

# Verify that the cue in play.ground/scripts:main evaluates as expected
exec cue export --out cue play.ground/scripts:main
cmp stdout cue.golden

# Build our main program
go build

# Verify that our main program gives the same output
exec ./play.ground
cmp stdout go.golden

# Remove the scripts directory to prove that we are not cheating
# by loading the CUE files from there
rm scripts
exec ./play.ground
cmp stdout go.golden

-- cue.mod/module.cue --
module: "play.ground"
-- go.mod --
module play.ground

go 1.16

require cuelang.org/go v0.4.0
-- main.go --
package main

import (
	"embed"
	"fmt"
	"io"
	"io/fs"
	"log"
	"os"
	"path/filepath"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

// Define our embed.FS.
//
//go:embed cue.mod/module.cue scripts/*.cue scripts/pipeline/*.cue
var staticFS embed.FS

func main() {
	// Loading staticFS via CUE we have two options:
	//
	// 1. Write staticFS to a temporary directory and
	//    call plain cue/load.Instances
	// 2. Create an overlay using the contents of staticFS
	//    and call cue/load.Instances with a config that
	//    uses the overlay and an empty directory as the
	//    working directory equivalent
	//
	// For this example we use option 2

	// Create a temporary directory so that the load of the overlay is
	// guaranteed to happen within an empty directory. An overlay complements
	// what is on disk, so we need to be sure that load doesn't pick up anything
	// other than the overlay.
	td, err := os.MkdirTemp("", "")
	if err != nil {
		log.Fatal(err)
	}
	defer os.RemoveAll(td)

	// Create the overlay. Note the overlay is a mapping of absolute filepaths
	// to contents. So we need to prepend the filepath from the embed.FS with
	// our temporary directory, the directory in which we are performing the
	// load.
	overlay := make(map[string]load.Source)
	err = fs.WalkDir(staticFS, ".", func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.Type().IsRegular() {
			return nil
		}
		f, err := staticFS.Open(p)
		if err != nil {
			return err
		}
		byts, err := io.ReadAll(f)
		if err != nil {
			return err
		}
		op := filepath.Join(td, p)
		overlay[op] = load.FromBytes(byts)
		return nil
	})
	if err != nil {
		log.Fatal(err)
	}

	// Load the build instances. Attempting to load play.ground/scripts:main
	// will succeed because Dir is set to the temporary directory, a directory
	// structure that is augmented by the overlay, and the overlay contains
	// cue.mod/module.cue, which defines the module path to be play.ground.
	// Notice that we need the :main package selector, because
	// play.ground/scripts does not identify the main package.
	config := &load.Config{
		Overlay: overlay,
		Dir:     td,
	}
	inst := load.Instances([]string{"play.ground/scripts:main"}, config)[0]
	if err := inst.Err; err != nil {
		log.Fatal(err)
	}
	ctx := cuecontext.New()
	v := ctx.BuildInstance(inst)
	fmt.Printf("%v\n", v)
}
-- scripts/main.cue --
package main

import (
	"play.ground/scripts/pipeline"
)

s: pipeline.#Schema & {
	Name: "jdoe"
	Age:  2000
	bebe: 11
}
-- scripts/pipeline/schema.cue --
package pipeline

#Schema: {
	Name: string
	Age:  >1000
	bebe: >1
}
-- cue.golden --
s: {
	Name: "jdoe"
	Age:  2000
	bebe: 11
}
-- go.golden --
{
	s: {
		Name: "jdoe"
		Age:  2000
		bebe: 11
	}
}
```

**Reply by yassinm** [reply]
@myitcv thank you so much for your detailed answer ! i really cant say how much i appreciate this !

Note: In my particular case i still cant use a temp directory since the main.cue file is something that is passed to my application from the cli. That is why i used cwd as the prefix for the overlay. I can see how users could pass a dir with files that will be loaded before the ones that ship with my binary. For now I am not sure i should stop that from happening since this could be seen as a nice feature :-) Meaning a user could be allowed to purposefully force the tool to load from the local file system instead of using the file that shipped with the binary for certain files.

**Reply by myitcv** [reply]
Ah I see, so `main.cue` is passed by the user; what about the contents of `scripts`?

**Reply by verdverm** [reply]
@yassinm things can get complicated when

1) users unintentionally override your values (hard to debug)
2) you want to ship a new schema and the tool at the same time (users may only want one of the updates)

short-term, you could read any user files and add them to the overlay for more precise control

long-term, dep management and fs.FS support should go a long way towards alleviating these issues

**Reply by myitcv** [reply]
> dep management

Exactly what came to mind, which is where #851 comes in.

**Reply by yassinm** [reply]
One aspect (read feature) i was thinking about specifically about embedding files in go. This could be a separate issue altogether but since we are talking about overlay/ modules here we go.

The system assumes the embed binary will load all the files in memory via the current overlay method. This is tedious and maybe wasteful. Is it possible to move to a world where overlay is not loaded upfront in the future ? This would make it easier when you have a good amount of files embedded in your go binary. For example a possible solution would be to craft an api based on a "resolver". 

1) A resolver can be registered for a particular "path". In this particular example it would be something like `cue.Register("play.ground/scripts", aResolver)` 
2) A resolver would basically be responsible to return a byte array when given a path like
```go
type (
	Resolver interface {
		Resolve(path string) ([]byte, error)
	}
)
```

3) if people can piggy back on this resolver API i can see how this could bypass the need for a module system specific for cue. Assuming someone can write a resolver which would be a few lines of go code. Then this would alleviate writing a cue oriented module system because go already has versioning and a module system support.

update:
3) is a orthogonal and not so important. I am more interested in the resolver based api to move away from overlay.

**Reply by yassinm** [reply]
> Ah I see, so `main.cue` is passed by the user; what about the contents of `scripts`?

Everything under scripts in my case is supposed to be embedded in the binary. I am still mulling if i should allow loading files under the "play.ground/scripts" path to be loaded from disk or no for certain cases. With the Resolver api i depicted above I would have more control and be more deterministic.

@verdverm I keep thinking that the Resolver api could also solve some of the problems you highlighted. because It gives more control.

**Reply by myitcv** [reply]
@yassinm I'm not entirely following your suggestion above for a resolve API. 

The current situation is that overlays are used to augment what is on disk.

We plan to add support for `cue/load` loading from an `io/fs.FS` instead of disk. This means that where files are can be entirely abstracted by the caller, to the extent that we could also do away with overlays.

Do you see any gaps in this proposal?

**Reply by yassinm** [reply]
@myitcv The approach i am proposing is to not be specific to any fs. If we are going to fix the API and do away with overlays maybe now is the time to consider another approach that is not specific to io/fs.FS.

1) The resolver API above is simple.
2) it can be adapted to any fs out there. Be it local from io/fs.FS.
3) Its basically based on a callback API. To me It gives the caller more control on what to allow to be loaded and from where. It it is not specific to io/fs.FS
4) It forces cue to be more deterministic. Meaning once a mount point like "play.ground/scripts" is mapped to a particular resolver it should only be resolved by that resolver or an error will be returned
5) it can be implemented now without waiting for io.FS to land in 1.17

**Reply by verdverm** [reply]
I'm not sure how a dynamic callback would force cue to be more deterministic. If such a callback had conditionals, could it not become non-deterministic. Or if the callback used an API (remote source) for example?

To me, the fs.FS would create a common interface for both the `cue` tool and API usage. My understanding is that the fs.FS is meant to be an interface that can be adapted by any fs out there. One would then be free to use a resolver interface to fill the FS prior to passing it to the load. In general, the FS interface seems to increase the determinism, as there would be no dynamic file loading after the load process is started. This is of course sidestepping how imported module loading happens, however I would expect that the `cue` cli would map the shared cache created by dep management in. Makes me wonder if the eventual FS mechanisms need to support the user supplying multiple FS that are merged or layered for lookups that may happen? I lack some details on how mapping local FS performs i.e. we wouldn't want to load all of the local mod cache on every call, only those necessary, but that would only be possible once the entrypoints have been loaded and imports understood. If a local FS mapping is lazy, then I would not expect a perf issue from mapping the entire local mod cache on each call (by the `cue` cli code path).

Go 1.17 is not that far out either.

**Reply by yassinm** [reply]
> I'm not sure how a dynamic callback would force cue to be more deterministic. If such a callback had conditionals, could it not become non-deterministic.

Right now the caller cannot decide the order in which file system resolution happens. Its somewhere in cue but it is not something that the caller can dictate. If we instead have multiple revolvers each "mounted" at a particular point then the caller could have more control on fs resolution. something like
```
cue.Register("play.ground/scripts", ioFsResolver)
cue.Register("play.ground/scripts/git", gitResolver)
cue.Register("play.ground/scripts/http", httpResolver)
cue.Register("/", defaultResolver)
```

1) There is nothing that says you cant have both methods living side by side.
2) If i had to chose however i can see that In the callback API we can make sure cue does not not care/know about the existence of a fs. That part is abstracted. So the question then becomes should cue become aware of the existence of a fs ? can/should that be abstracted ? what do we lose if remove cue dependence on the existence of a fs ? what do we gain by making it aware of a fs ?

My take is it should not but i could be wrong or missing some additional context :-) I am not saying you cant make it work behind io/fs.FS. I am asking if it should be via that interface or a resolver interface ?

**Reply by verdverm** [reply]
There are directories and files either way, so the "FS" abstraction seems unavoidable. How are modules and packages to be organized otherwise?

Again the Golang FS package is an abstraction (set of interfaces) that is not tied to an OS filesystem. It could be disk, backed, memory based, or some sort of resolver based system that you are describing. You could easily implement a resolver based system using this interface and pass that fs.FS type to the CUE API

**Reply by myitcv** [reply]
@yassinm I'm still not entirely clear why `io/fs.FS` cannot be that abstraction you are talking about. It is, in effect, almost exactly the "callback" you are defining above. 

```
type FS interface {
	// Open opens the named file.
	//
	// When Open returns an error, it should be of type *PathError
	// with the Op field set to "open", the Path field set to name,
	// and the Err field describing the problem.
	//
	// Open should reject attempts to open names that do not satisfy
	// ValidPath(name), returning a *PathError with Err set to
	// ErrInvalid or ErrNotExist.
	Open(name string) (File, error)
}
```

> Right now the caller cannot decide the order in which file system resolution happens

Is there a reason this is significant? 

> I am asking if it should be via that interface or a resolver interface ?

I think we would be willing to consider an alternative if there are problems with `io/fs.FS`, or clear benefits of the alternative. `io/fs.FS` is just an interface. Whether it maps to a real file system, something in memory, a hybrid, something HTTP-based... is entirely undefined and left up to implementations. 

The one clear benefit of using `io/fs.FS` to my mind is that it limits the API surface of the CUE Go API, by reusing a well defined, well known concept from the Go standard library. The cognitive load is therefore lower etc.

**Reply by yassinm** [reply]
@myitcv that makes a lot of sense thank you :-)

I think the only reservation i have left is about the order of the resolution and restricting one FS to a prefix path. In my use case i need to make sure a particular path like "play.ground/scripts/*" is only resolved by my fs implementation and not loaded from the disk. I can move that discussion to where the FS interface implementation is happening which makes more sense.

**Reply by myitcv** [reply]
> In my use case i need to make sure a particular path like "play.ground/scripts/*" is only resolved by my fs implementation and not loaded from the disk.

And that's entirely doable with `io/fs.FS`. Because you pass the `io/fs.FS` value to `cue/load.Config` and hence you control the implementation.

**Reply by myitcv** [reply]
For anyone following along here, we have just created a proposal for `io/fs.FS` support within `cue/load`: https://github.com/cue-lang/cue/discussions/4285
